### PR TITLE
Add release workflow

### PR DESCRIPTION
**Pull request by LordAro**
Not finished yet

* [x] Build Python wheels for Windows, MacOS & Linux (manylinux2010)
* [x] Builds standalone (win64) nmlc executable
* [x] Fixes a bug in git version calculation with tags
* [x] Make py38 work ( RalfG/python-wheels-manylinux-build#14 )
* [x] Work out which artefacts want to be distributed with the release (fixes Linux release failure)
* [x] ~See if build artefacts can be attached to the release~ Not worth the bother
* [x] Remove test code
* [x] Add `PYPI_PASSWORD` secrets to repo

Latest release build example: https://github.com/LordAro/nml/actions/runs/83157723

### CODE REVIEW

File: .github/workflows/regression.yml [hunk 1]
```diff
@@ -64,3 +64,10 @@ jobs:
       shell: bash
       env:
         NMLC: nmlc
+
+    # Just check we can successfully build it
```
Comment (TrueBrain) [inline .github/workflows/regression.yml:68]: for other projects, we do this in "testing.yml", as there might be other things to test for (like flake, black, ..)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
+name: Release
+
+
+on:
+  release:
+    types: [created]
+
+jobs:
+  regression:
+
```
Comment (TrueBrain) [inline .github/workflows/release.yml:10]: silly newline

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
+name: Release
+
+
```
Comment (TrueBrain) [inline .github/workflows/release.yml:3]: silly newline

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+on:
+  release:
+    types: [created]
```
Comment (TrueBrain) [inline .github/workflows/release.yml:6]: In other projects we do this on "push/tags". Not sure if they differ. Might be good to either change them all to the one or the other.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        twine upload --repository-url https://test.pypi.org/legacy/ dist/*.whl
+
+    - name: Build standalone executable
+      if: startsWith(matrix.os, 'windows')
+      run: |
+        pip install --upgrade pyinstaller
+        pyinstaller ./nmlc.spec
+        $version = python ./setup.py --version
+        echo "::set-output name=version::$version"
+      id: exe
```
Comment (TrueBrain) [inline .github/workflows/release.yml:87]: I am sure this id can be better.

File: nml/version_info.py [hunk 6]
```diff
@@ -119,8 +119,8 @@ def get_git_version(detailed = False):
         # Compose the actual version string
         str_tag = ""
         if len(tag) > 0:
```
Comment (TrueBrain) [inline nml/version_info.py:121]: while at it, fix this line too: `if tag:` :)

File: setup.py [hunk 7]
```diff
@@ -17,7 +17,7 @@ def run(self):
         super().run()
 
 setup(
-    name='nml',
+    name='nml-test',
```
Comment (TrueBrain) [inline setup.py:20]: temporary, I assume?

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      with:
+        path: ~\AppData\Local\pip\Cache
+        key: ${{ runner.os }}-pip-${{ hashFiles('**/setup.py') }}
+        restore-keys: |
+          ${{ runner.os }}-pip-
+
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: '3.x'
```
Comment (glx22) [inline .github/workflows/release.yml:48]: Regression uses `architecture: ${{ startsWith(matrix.os, 'windows') && 'x86' || 'x64' }}`, release defaults to `architecture: 'x64'` as it's not specified.

Comment (planetmaker) [reply .github/workflows/release.yml:48]: it needs python >= 3.5 at least

File: .github/workflows/regression.yml (hunk 1 above)
Comment (LordAro) [reply .github/workflows/regression.yml:68]: out of scope for this PR, I feel :p

File: .github/workflows/release.yml (hunk 8 above)
Comment (LordAro) [reply .github/workflows/release.yml:48]: "3.x" is just "the latest", afaik

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,101 @@
+name: Release
+
+
+on:
+  release:
+    types: [created]
+
+jobs:
+  regression:
```
Comment (TrueBrain) [inline .github/workflows/release.yml:9]: too much copy/pasting :D

File: setup.py (hunk 7 above)
Comment (LordAro) [reply setup.py:20]: From the PR description
> * [ ] Remove test code

Yes :p Necessary for testing upload to test.pypi.org, as nml already exists (surprisingly enough)

File: .github/workflows/release.yml (hunk 4 above)
Comment (TrueBrain) [reply .github/workflows/release.yml:5]: The only thing I prefer is that we do the same thing for all projects. So pick one and apply it everywhere. This is not about personal preferences, this is about keeping things sane :)

**Comment by MinchinWeb** [comment]
Oooh... does this mean we'll start seeing "regular" PyPI releases?

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+on:
+  push:
+    branches:
+    - master
+    tags:
+    - '*'
+  repository_dispatch:
+    types:
+    - publish_latest_tag
+    - publish_master
```
Comment (TrueBrain) [inline .github/workflows/release.yml:12]: are you using the dispatch?

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        python setup.py sdist bdist_wheel
+        twine upload --username __token__ dist/*.whl
+
+    # Only do this on Windows to prevent file conflicts
+    - name: Build and publish source distribution
+      if: startsWith(matrix.os, 'windows')
+      env:
+        TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+      run: |
+        twine upload --username __token__ --repository-url https://test.pypi.org/legacy/ dist/*.tar.gz
```
Comment (TrueBrain) [inline .github/workflows/release.yml]: this repository url intended?

File: .github/workflows/testing.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        pip install -e .
+
+    # Just check we can successfully build it
+    - name: Build standalone executable
+      if: startsWith(matrix.os, 'windows')
+      run: |
+        pip install pyinstaller
+        pyinstaller ./nmlc.spec
+
+    # TODO: Add black/flake8 tests
```
Comment (TrueBrain) [inline .github/workflows/testing.yml:62]: For another PR?

Comment (LordAro) [reply .github/workflows/testing.yml:62]: That was the intent (will require quite a few code changes)

Comment (TrueBrain) [reply .github/workflows/testing.yml:62]: I meant: please remove the comment and do this in another PR :P

File: .github/workflows/release.yml (hunk 10 above)
Comment (LordAro) [reply .github/workflows/release.yml:12]: Not sure what you mean? All I know is that works on my repo (see release link in main body)

**Comment by andythenorth** [comment]
Merging, as TB has approved.

**Pull request merged**